### PR TITLE
Add options to deny editing of Ident, RealName, QuitMsg, Networks/Servers & CTCP Replies

### LINKED ISSUE #1682: [Feature Request] Add ability do disallow normal users to change their ident, add/remove/edit networks/servers and quitmsg
After a brief discussion with MetaNova in IRC and advised by him, i decided to open this feature request.

This feature was implemented by PanicBNC [here](https://github.com/PanicBNC-Dev/ZNC) with a special flag named `Premium`.

##### Small IRC excerpt:
```
[00:10:22] <EliteGod> I'm heading to bed. I'm just leaving a small request here: I've used a PanicBNC custom build for some time. in that build normal users couldn't change their ident, add/remove networks/servers (via IRC or webpanel). is there anyone interested in adapting those changes to znc 1.7.4 in exchange of some sort of Patreon or something like that?
[00:11:26] <MetaNova> if you open a ticket (feature request) I'll put money towards a bounty on it
[00:12:31] <EliteGod> I'll do it tomorrow for sure. do you need the PanicBNC build (they have a GitHub one) to see the changes they've made?
[00:13:04] <EliteGod> it would be just awesome for people that want to provide a public ZNC service
[00:13:08] <MetaNova> relevant links to examples of the implementation elsewhere should be included in the ticket
[00:13:20] <MetaNova> for reference
[00:14:25] <MetaNova> and, yes, I agree: having the ability to prevent users from changing various aspects about their user (ident response, realname, quitmsg, and even the ability to add/remove networks/servers) be part of native ZNC would simplify a good many providers' lives
```

As stated by MetaNova this would help providers and simplify their lives, as probably most of users that run ZNC personally have only one user with all the networks they need.

##### What i think that should be blocked by default:
Deny SetBindHost (checked by default)
Ability to change the ident
Ability to add/remove/edit servers/networks
Ability to change realname
Ability to change quitmsg

All of this should be blocked via IRC commands/webpanel.

@magnum7385 
Feel free to put the bounty on this and add your ideas.

PS: Sorry if my English is bad but i'm Portuguese

### PR DESCRIPTION
**Pull request by BradleyShaw**
Closes #1682 

Add the following options:
* `DenySetIdent` - Denies setting ident
* `DenySetNetwork` - Denies adding/removing networks/servers
* `DenySetRealName` - Denies setting realname
* `DenySetQuitMsg` - Denies setting quitmsg
* `DenySetCTCPReplies` - Denies adding/removing CTCP replies

### CODE REVIEW

**Comment by mtnz0r** [comment]
incrediblenews

**Comment by loop-de-r00t** [comment]
Just tried this branch on a throwaway VPS and it works nicely. Awesome job

File: modules/data/webadmin/tmpl/add_edit_network.tmpl [hunk 1]
```diff
@@ -21,15 +25,21 @@
 		<input type="hidden" name="user" value="<? VAR Username ?>" />
 		<input type="hidden" name="network" value="<? VAR Name ?>" />
 
+		<? IF !NameEdit ?>
+		<input type="hidden" name="name" value="<? VAR Name ?>" />
```
Comment (DarthGandalf) [inline modules/data/webadmin/tmpl/add_edit_network.tmpl:29]: This value already exists just above, why do you need another one?

File: modules/data/webadmin/tmpl/add_edit_network.tmpl [hunk 2]
```diff
@@ -42,16 +52,20 @@
...
 				</div>
 				<div style="clear: both;"></div>
+				<? IF IdentEdit ?>
 				<div class="subsection">
 					<div class="inputlabel"><label for="ident"><? FORMAT "Ident:" ?></label></div>
 					<input id="ident" type="text" name="ident" value="<? VAR Ident ?>" class="half" maxlength="128"
 						   title="<? FORMAT "Your ident." ?>"/>
 				</div>
+				<? ENDIF ?>
+				<? IF RealNameEdit ?>
```
Comment (DarthGandalf) [inline modules/data/webadmin/tmpl/add_edit_network.tmpl:62]: Maybe still show the field, just make it readonly?

File: modules/webadmin.cpp [hunk 3]
```diff
@@ -1145,10 +1221,12 @@ class CWebAdminMod : public CModule {
 
         VCString vsArgs;
 
-        pNetwork->DelServers();
-        WebSock.GetRawParam("servers").Split("\n", vsArgs);
-        for (const CString& sServer : vsArgs) {
-            pNetwork->AddServer(sServer.Trim_n());
+        if (spSession->IsAdmin() || !spSession->GetUser()->DenySetNetwork()) {
```
Comment (DarthGandalf) [inline modules/webadmin.cpp:1224]: This DenySetNetwork looks like a catch-all for all network-level settings, *and* adding new networks. What exactly it's supposed to achieve?

File: modules/webadmin.cpp [hunk 4]
```diff
@@ -1573,15 +1675,50 @@ class CWebAdminMod : public CModule {
...
+                CTemplate& o14 = Tmpl.AddRow("OptionLoop");
+                o14["Name"] = "denysetrealname";
+                o14["DisplayName"] = t_s("Deny SetRealName");
+                if (pUser->DenySetRealName()) {
+                    o14["Checked"] = "true";
+                }
+
+                CTemplate& o15 = Tmpl.AddRow("OptionLoop");
+                o15["Name"] = "denysetquitmsg";
+                o15["DisplayName"] = t_s("Deny SetQuitMsg");
```
Comment (DarthGandalf) [inline modules/webadmin.cpp:1702]: Can you make the descriptions more descriptive?

**Comment by DarthGandalf** [comment]
Can you add any tests?

File: modules/webadmin.cpp (hunk 3 above)
Comment (BradleyShaw) [reply modules/webadmin.cpp:1224]: It prevents adding and deleting networks, changing the network name and modifying the server list in any way

File: modules/data/webadmin/tmpl/add_edit_network.tmpl (hunk 1 above)
Comment (BradleyShaw) [reply modules/data/webadmin/tmpl/add_edit_network.tmpl:29]: It does appear to be unnecessary, I'll remove it

**Comment by BradleyShaw** [comment]
I've made the changes suggested in review, I'm not sure how to add tests

**Comment by DarthGandalf** [comment]
Take a look at test/integration/test directory

**Comment by BradleyShaw** [comment]
Added tests for all DenySet* options

**Comment by DarthGandalf** [comment]
> will decrease coverage by 5.37%.

Ignore this. It accidentally only measures perl coverage right now, after switch from Travis several days ago, ignoring any C++. I'm trying to fix it at #1798

**Comment by DarthGandalf** [comment]
Выделение_379

Could this show a better error message than a weird gray rectangle?

**Comment by DarthGandalf** [comment]
When JS is disabled, CTCP/Server editors are still editable. Also I think "disabled" is probably better than "readonly" for text fields, because it's more obvious that it's not editable

**Comment by bashgeek** [comment]
@BradleyShaw I apologise for jumping in here unannounced, but are you still working on this or could someone else take this over to get this to the finish line? We'd be willing to take this on if nobody objects, seems only small cosmetic changes are required to get this done. Thanks!

**Pull request closed without merging**